### PR TITLE
Generate a default closure in InterceptingGCL

### PR DESCRIPTION
**Pull request by nestoracunablanco**
In case a method is matched with null value closure, the default generated closure is {}.

This does not work when the method has more parameters, so the approach of generating a default closure using dynamic compilation was taken.

- [X] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [X] Ensure that the pull request title represents the desired changelog entry
- [X] Please describe what you did
- [X] Link to relevant issues in GitHub or Jira
- [X] Link to relevant pull requests, esp. upstream and downstream changes
- [X] Ensure you have provided tests - that demonstrates feature works or fixes the issue

Solves comments in issue #461

### CODE REVIEW

File: src/main/groovy/com/lesfurets/jenkins/unit/InterceptingGCL.groovy [hunk 1]
```diff
@@ -18,11 +18,20 @@ class InterceptingGCL extends GroovyClassLoader {
...
                 // a matching method was registered, replace script method execution call with the registered closure (mock)
-                metaClazz."$scriptMethod.name" = matchingMethod.value ?: {}
+                metaClazz."$scriptMethod.name" = matchingMethod.value ?: defaultClosure(matchingMethod.key.args)
             }
         }
     }
 
+    static Closure defaultClosure(Class[] args) {
+        String argumentsString = args.inject("") { acc, value ->
+            acc + value.name + " " + "a"*(acc.count(',') + 1) + ","}
```
Comment (nre-ableton) [inline src/main/groovy/com/lesfurets/jenkins/unit/InterceptingGCL.groovy:28]: IMHO this is more readable:

```suggestion
        String argumentsString = args.inject("") { acc, value ->
            return "${acc}${value.name} a${*(acc.count(',') + 1)},"
        }
```

File: src/main/groovy/com/lesfurets/jenkins/unit/InterceptingGCL.groovy [hunk 2]
```diff
@@ -18,11 +18,20 @@ class InterceptingGCL extends GroovyClassLoader {
...
+        String argumentsStringWithoutComma = argumentsString.size() > 0 ?
+                argumentsString.substring(0, argumentsString.length()-1) : argumentsString
+        String closureString = "{$argumentsStringWithoutComma -> }"
+        return (Closure)new GroovyShell().evaluate(closureString)
```
Comment (nre-ableton) [inline src/main/groovy/com/lesfurets/jenkins/unit/InterceptingGCL.groovy:32]: I don't think a cast is the way to go here. Why not return a real closure?

```suggestion
        return { new GroovyShell().evaluate(closureString) }
```

**Review by nre-ableton — CHANGES_REQUESTED** [review]
There is a lot of logic going on in `defaultClosure` that isn't tested. It would be nice to have a unit test to make sure that all of the string slicing and appending works as expected.

Also, there is a typo in the commit message (it should be "_Generate_ a default ..."). I fixed this in the PR title for you.

File: src/main/groovy/com/lesfurets/jenkins/unit/InterceptingGCL.groovy (hunk 1 above)
Comment (nestoracunablanco) [reply src/main/groovy/com/lesfurets/jenkins/unit/InterceptingGCL.groovy]: This change did not work out of the box, instead I did something similar. I hope I got the idea behind your suggestion.

File: src/main/groovy/com/lesfurets/jenkins/unit/InterceptingGCL.groovy (hunk 2 above)
Comment (nestoracunablanco) [reply src/main/groovy/com/lesfurets/jenkins/unit/InterceptingGCL.groovy:34]: I agree with you, that would be nicer... but this did not work. Testing the suggested way, it created a closure with a closure inside, which is not intended.

**Comment by nestoracunablanco** [comment]
@nre-ableton more unit tests are now in place.

File: src/main/groovy/com/lesfurets/jenkins/unit/InterceptingGCL.groovy [hunk 3]
```diff
@@ -18,11 +18,22 @@ class InterceptingGCL extends GroovyClassLoader {
...
+                metaClazz."$scriptMethod.name" = matchingMethod.value ?: defaultClosure(matchingMethod.key.args)
             }
         }
     }
 
+    static Closure defaultClosure(Class[] args) {
+        int maxLength = 254
+        if (args.length > maxLength) { throw new IllegalArgumentException("Only $maxLength arguments allowed")}
+        String argumentsString = args.inject("") { acc, value ->
+            return "${acc}${value.name} ${'a'*(acc.count(',') + 1)},"}
```
Comment (nre-ableton) [inline src/main/groovy/com/lesfurets/jenkins/unit/InterceptingGCL.groovy:30]:
```suggestion
        if (args.length > maxLength) {
            throw new IllegalArgumentException("Only $maxLength arguments allowed")
        }
        String argumentsString = args.inject("") { acc, value ->
            return "${acc}${value.name} ${'a'*(acc.count(',') + 1)},"
        }
```

...just a minor suggestion to improve readability.

**Review by nre-ableton — APPROVED** [review]
I guess this change looks fine to me, but I am not sure I can speak to the broader impact that this change may have. It would be nice if @stchar could also take a look.

**Comment by nestoracunablanco** [comment]
Hi @stchar, can you please also take a look at this pull request? Thanks!

**Review by stchar — APPROVED** [review]
Awesome!

**Comment by stchar** [comment]
@nestoracunablanco, @nre-ableton, reviewed. Looks good. 
I would also add tests for scripted pipeline, but I assume there is no any difference as `InterceptingGCL` address to metaclass of `groovy.lang.Script` which is the same in case of Declarative and Scripted pipeline test runs.

**Comment by nestoracunablanco** [comment]
Thanks for the nice feedback!

**Pull request merged**